mttl/config.py

```diff
@@ -133,229 +197,321 @@ def save_config(self, output_dir):
-        self.finetune_use_last_checkpoint = (
-            False  # use always the best valid_perf checkpoint if available
-        )
+        if cmd_kwargs:
+            overwrite_logs += cls.process_kwargs(
+                cmd_kwargs,
+                raise_error=raise_error,
+            )
+            extra_kwargs.update(cmd_kwargs)
+
+        # log the overwrites
+        for log in overwrite_logs:
+            logger.warning(log)
 
-        self.model = None
-        self.model_family = None  # model family, either "gpt" or "encdec"
-        self.attn_implementation = None
+        config = cls(**extra_kwargs)
 
-        self.precision = "32"
-        self.monitor_grad_alignment_on = None
+        if return_parser:
+            return config, args
+        return config
 
-        self.model_modifier = None
-        self.adapter_type = None
 
-        self.lora_rank = 16
-        self.lora_dropout = 0.0
-        self.lora_init_scale = 0.01
-        self.lora_alpha = 1.0
-        self.lora_warmup = False
-        self.lora_init_b_random = False
-        self.lora_dropout = 0.0
+class MetaRegistrable(type):
+    """
+    Meta class that creates a new dataclass containing all the configs
+    in this registrable.
+    """
+
+    def __new__(cls, name, bases, attrs, registrable: Registrable = None):
+        module_name, class_name = registrable.rsplit(".", 1)
+        module = importlib.import_module(module_name)
+
+        registrable_class = getattr(module, class_name)
+
+        # make the union of all the fields across the registered configs
+        to_tuples = dataclasses_union(*registrable_class.registered_configs())
+
+        # create new dataclass with the union of all the fields
+        new_cls = make_dataclass(name, to_tuples, bases=(Args,), init=False)
+        new_cls.registrable_class = registrable_class
+
+        # set functions to be had in the new baby dataclass
+        for k, v in {
+            **attrs,
+        }.items():
+            setattr(new_cls, k, v)
+        return new_cls
+
+
+@dataclass
+class DataArgs(
+    metaclass=MetaRegistrable, registrable="mttl.datamodule.base.DataModule"
+):
+    pass
+
+
+@dataclass
+class SelectorArgs(
+    metaclass=MetaRegistrable,
+    registrable="mttl.models.containers.selectors.base.Selector",
+):
+    pass
+
+
+@dataclass
+class ModifierArgs(
+    metaclass=MetaRegistrable, registrable="mttl.models.modifiers.base.Modifier"
+):
+    pass
+
+
+@dataclass
+class TransformArgs(
+    metaclass=MetaRegistrable,
+    registrable="mttl.models.library.library_transforms.LibraryTransform",
+):
+    pass
+
+
+@dataclass
+class TrainingArgs(DataArgs):
+    cache_dir: str = os.getenv("CACHE_DIR", "./cache")
+    data_dir: str = os.getenv("TRAIN_DIR", "/tmp/")
+    output_dir: str = os.getenv("OUTPUT_DIR", "./output")
+
+    finetune_task_name: str = None
+    exp_name: str = None
+    expert_name: str = None
+
+    # Training config
+    micro_batch_size: str = None
```

Comment:
Now would be a good time to clean up the mess that is `micro_batch_size` `train_batch_size` `gradient_accumulation_steps`.

Reply:
oh, do we have the three of them?

Reply:
alloa

Reply:
`gradient_accumulation_steps` is computed in post_init so it's not settable by the user, it's fine no?